src/curve_intersections.rs

```diff
@@ -0,0 +1,1377 @@
+        const NONE                                  = 0b00000000;
+        const KEEP_CURVE1_T0_INTERSECTION           = 0b00000001;
+        const KEEP_CURVE1_T1_INTERSECTION           = 0b00000010;
+        const KEEP_CURVE2_T0_INTERSECTION           = 0b00000100;
+        const KEEP_CURVE2_T1_INTERSECTION           = 0b00001000;
+        const KEEP_CURVE1_ENDPOINT_INTERSECTIONS    = Self::KEEP_CURVE1_T0_INTERSECTION.bits
+                                                    | Self::KEEP_CURVE1_T1_INTERSECTION.bits;
+        const KEEP_CURVE2_ENDPOINT_INTERSECTIONS    = Self::KEEP_CURVE2_T0_INTERSECTION.bits
+                                                    | Self::KEEP_CURVE2_T1_INTERSECTION.bits;
+        const KEEP_ALL_ENDPOINT_INTERSECTIONS       = Self::KEEP_CURVE1_ENDPOINT_INTERSECTIONS.bits
+                                                    | Self::KEEP_CURVE2_ENDPOINT_INTERSECTIONS.bits;
+        const KEEP_DUPLICATE_INTERSECTIONS          = 0b00010000;
+    }
+}
+
+/// Compute the intersections between two Bézier curves
+pub fn curve_curve_intersections<T: ParamCurveBezierClipping, U: ParamCurveBezierClipping>(
+    curve1: &T,
+    curve2: &U,
+    flags: CurveIntersectionFlags,
+    accuracy: f64,
+) -> ArrayVec<(f64, f64), 9> {
+    let mut av = ArrayVec::new();
+
+    // Make sure we don't use too small of an accuracy
+    let bounds = curve1.bounding_box().union(curve2.bounding_box());
+    let accuracy = Point::epsilon(Point::new(bounds.max_x(), bounds.max_y()), accuracy);
+
+    // Check if both curves lie directly on each other for some span
+    let overlaps = check_for_overlap(curve1, curve2, flags, accuracy, &mut av);
+    if overlaps {
+        return av;
+    }
+
+    // Find intersections
+    add_curve_intersections(
+        curve1,
+        curve2,
+        &(0.0..1.0),
+        &(0.0..1.0),
+        &mut av,
+        false,
+        0,
+        0,
+        curve1,
+        curve2,
+        flags,
+        accuracy,
+    );
+    av
+}
+
+/// This function tests if one curve lies along another curve for some range
+fn check_for_overlap<T: ParamCurveBezierClipping, U: ParamCurveBezierClipping>(
+    curve1: &T,
+    curve2: &U,
+    flags: CurveIntersectionFlags,
+    accuracy: f64,
+    intersections: &mut ArrayVec<(f64, f64), 9>,
+) -> bool {
+    // If the two curves overlap, there are two main cases we need to account for:
+    //   1) One curve is completely along some portion of the other curve, thus
+    //      both endpoints of the first curve lie along the other curve.
+    //      ex. curve 1: O------------------O
+    //          curve 2:    O---O
+    //   2) The two curves overlap, thus one endpoint from each curve lies
+    //      along the other.
+    //      ex. curve 1: O------------------O
+    //          curve 2:               O-----------O
+
+    // Use this function to get the t-value on another curve that corresponds to the t-value on this curve
+    #[inline]
```

Comment:
inline seems dubious here given the size of the function. I'd just let the compiler do its thing.